daffodil-runtime1/src/main/scala/org/apache/daffodil/layers/IPv4Checksum.scala

```diff
@@ -0,0 +1,403 @@
+  }
+  srd.schemaDefinitionUnless(finalChecksumVRD.primType == PrimType.UnsignedShort,
+    "Variable '%s' is not of type 'xs:unsignedShort'.", part1ChecksumVarQName.toExtendedSyntax)
+
+  /**
+   * Common code used for part1 and part2.
+   * @param state The parser state
+   * @param optOriginalStream The original InputStream that this layer lives on top of.
+   * @param partByteArray An array of bytes sized to hold the bytes of the part
+   * @param shortBuf A ShortBuffer pre-set to use the partByteArray as source of Short values in network byte
+   *                 order (aka 'bigEndian')
+   * @param partLengthInBytes Length of the partByteArray (which is the part length)
+   * @param initialChecksumValue Starting value for computation of this checksum, that is, an
+   *                             earlier partial checksum, or 0 if no previous.
+   * @return The checksum value for this part as an Int (32 bits). The checksum is a 16-bit
+   *         quantity, but the addition can produce carries that occupy more than 16 bits.
+   *         This return value has not yet added in any overflow bits from the most-significant 16-bits
+   *         into the least-significant 16-bits.
+   */
+  private def startShared(
+    state: ParseOrUnparseState,
+    optOriginalStream: Maybe[InputStream],
+    partByteArray: Array[Byte],
+    shortBuf: ShortBuffer,
+    partLengthInBytes: Int,
+    initialChecksumValue: Int): Int = {
+    //
+    // For parsing, all the hard work happens here, allowing the layered input stream to
+    // just deliver the bytes
+    //
+    Assert.invariant(state.bitPos0b % 8 == 0) // we are byte aligned.
+    Assert.invariant(optOriginalStream.isDefined)
+    val bitLen = partLengthInBytes * 8
+
+    try
+      IOUtils.readFully(optOriginalStream.get, partByteArray)
+    catch {
+      case eof: EOFException =>
+        throw DataInputStream.NotEnoughDataException(bitLen)
+    }
+    computeChecksum(initialChecksumValue, shortBuf, partLengthInBytes)
+  }
+
+  private def computeChecksum(initialChecksumValue: Int, shortBuf: ShortBuffer, partLengthInBytes: Int): Int = {
+    var i = 0
+    var partChecksum: Int = initialChecksumValue
+    val partNShorts = partLengthInBytes / 2
+    while (i < partNShorts) {
+      partChecksum += UShort(shortBuf.get(i)).toInt
+      Assert.invariant(partChecksum >= 0)
+      i += 1
+    }
+    //
+    // now combine the carry bits in the most significant 16 bits into the lower 16 bits.
+    //
+    val checksumLow = partChecksum & 0xFFFF
+    val checksumHigh = partChecksum >>> 16
+    val checksumTotal: Int = checksumLow + checksumHigh
+    Assert.invariant(checksumTotal <= 0xFFFF && checksumTotal >= 0)
+    val checksumTotalShort = UShort(checksumTotal.toShort)
+
+    checksumTotalShort.toInt
+  }
+
+  object Part1 { def apply() = new Part1() }
+
+  final class Part1 private ()
+    extends LayerTransformer() {
+
+    private val hdrPart1ByteArr = new Array[Byte](ipHdrPart1LengthInBytes)
```

Comment:
Add scaladoc explaining these members as their roles and the relationship between them is subtle.